### PR TITLE
Fixed an issue lambda deploy-serverless command tries to build image when ImageUri is set in template without any Metadata.

### PR DESCRIPTION
**Pull request by ashishdhingra**
*Issue #, if available:* #264

*Description of changes:*
Fixed an issue lambda deploy-serverless command tries to build image when ImageUri is set in template without any Metadata.

Per https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-resource-function.html#sam-function-imageuri, `Building your application with necessary Metadata entries takes precedence over ImageUri, so if you specify both then ImageUri is ignored.`.

In customer reported scenario, metadata is not present in YAML (orJSON) template. In such case, it should not try to build the Docker image locally, instead should use the `ImageUri` set in the CloudFormation template. Currently, before this fix, it tries to validate the `ImageUri` as local path and fails the validation check.

**NOTE:** We could have added a validation check to verify if image pointed out by the `ImageUri` exists in ECR, however, there is no API to get image by `ImageUri`.

___
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/Amazon.Lambda.Tools/TemplateProcessor/TemplateProcessorManager.cs [hunk 1]
```diff
@@ -142,12 +142,18 @@ private async Task<UpdateResourceResults> ProcessUpdatableResourceAsync(string t
             UpdateResourceResults results;
             var localPath = field.GetLocalPath();
 
-            if (!Path.IsPathRooted(localPath))
+            if (!field.IsImageUri && !Path.IsPathRooted(localPath))
```
Comment (normj) [inline src/Amazon.Lambda.Tools/TemplateProcessor/TemplateProcessorManager.cs:145]: We don't know if there were users that were relying on the .NET CLI building images with just the `ImageUri` property set. 

We can look to see if the value for ImageUri contains `.dkr.ecr.` and the first token is a numeric. If so then we can assume it is an ECR image and use that otherwise run the local build.

**Review by normj — CHANGES_REQUESTED** [review]
We are going to need a test that we can use a template that has an image uri set to an already pushed image.

File: src/Amazon.Lambda.Tools/TemplateProcessor/TemplateProcessorManager.cs (hunk 1 above)
Comment (ashishdhingra) [reply src/Amazon.Lambda.Tools/TemplateProcessor/TemplateProcessorManager.cs:145]: Added extra check in `IsCode` and `IsImageUri` properties in `UpdatableResource.cs`. Please review.

**Comment by ashishdhingra** [comment]
> We are going to need a test that we can use a template that has an image uri set to an already pushed image.

@normj Added tests for YAML and JSON template.

Could you please advise on the following:
- Possible 2 ways of specifying `ImageUri`:
  - The link https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-resource-function.html#sam-function-imageuri mentions `ImageUri` (and `CodeUri`) as top level template properties.
  - Whereas, https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-function.html specifies `ImageUri` nested inside top level code `Code` property. There is no `CodeUri` property.

  Which one is correct? Or both are supported? If yes, why 2 different ways of specifying `ImageUri`. Should we also test with nested `ImageUri` property template as well?

- During testing, I noticed that the cleanup `DeleteServerlessCommand` when executed without specifying the `Region` (in this case it would use the default region from the profile), even though was successful, but it was not deleting CloudFormation stack. The reason is that we use `TEST_REGION` (`sa-east-1`) while creation of stack and other resources, whereas the `DeleteServerlessCommand` was picking up some other region from my profile. Changing cleanup code to explicitly set `Region` as `TEST_REGION` actually initiated the stack deletion. This cleanup code was copied from existing test. It's highly likely that such existing test(s) might not be actually cleaning (deleting) up the stack. Should we review such cleanup code and correct it as part of this PR?

Thanks,
Ashish

File: src/Amazon.Lambda.Tools/TemplateProcessor/ITemplateParser.cs [hunk 2]
```diff
@@ -91,8 +91,13 @@ public interface IUpdateResourceField
         /// True if the field should contain code like a Lambda package bundle.
         /// </summary>
         bool IsCode { get; }
-        
-        
+
+        /// <summary>
+        /// True if the field points to an ImageUri.
+        /// </summary>
+        bool IsImageUri { get; }
```
Comment (normj) [inline src/Amazon.Lambda.Tools/TemplateProcessor/ITemplateParser.cs:98]: This name keeps confusing me since it is also matching the name of the property. My instinct is to think the property means the `ImageUri` property is set. What it really means is the `ImageUri` is pointing to already pushed image. Can we rename it to something like `IsImagePushed`?

File: src/Amazon.Lambda.Tools/TemplateProcessor/UpdatableResource.cs [hunk 3]
```diff
@@ -140,13 +141,44 @@ public bool IsCode
...
-                    if(!string.IsNullOrEmpty(this._resource.DataSource.GetValue("Code", "ZipFile")))
+                    if (!string.IsNullOrEmpty(this._resource.DataSource.GetValue("Code", "ZipFile")))
                     {
                         // The template contains embedded code.
                         return false;
-                    }                     
+                    }
+                    else
+                    {
+                        string localPath = GetLocalPath();
```
Comment (normj) [inline src/Amazon.Lambda.Tools/TemplateProcessor/UpdatableResource.cs:151]: I find this logic hard to follow. In particular comparing `localPath` with `tempLocalPath` and what is a `tempLocalPath`. What about something like this.

```
string localPath = this._resource.DataSource.GetValueFromResource(LambdaConstants.CF_SERVERLESS_METADATA, LambdaConstants.CF_SERVERLESS_DOCKERCONTEXT);
// Pointing to already pushed image, no package needs to be done.
if (IsECRImage(localPath))
{
    return false;
}
// metadata points to local path that needs to be packaged up return true for code.
else if (!string.IsNullOrEmpty(localPath))
{
    return true;
}

// If no docker metadata was set fallback to looking for a local path in the resource's ImageUri property.
localPath = this._resource.DataSource.GetValue(LambdaConstants.CF_LAMBDA_IMAGEURI);
// Pointing to already pushed image, no package needs to be done.
if (IsECRImage(localPath))
{
    return false;
}

return true;
```

File: src/Amazon.Lambda.Tools/TemplateProcessor/UpdatableResource.cs [hunk 4]
```diff
@@ -140,13 +141,44 @@ public bool IsCode
...
 
-                    return true;
+            public bool IsImageUri
+            {
+                get
+                {
+                    string localPath = GetLocalPath();
+
+                    string tempLocalPath = this._resource.DataSource.GetValueFromResource(LambdaConstants.CF_SERVERLESS_METADATA, LambdaConstants.CF_SERVERLESS_DOCKERCONTEXT);
+                    if (string.IsNullOrEmpty(tempLocalPath))
```
Comment (normj) [inline src/Amazon.Lambda.Tools/TemplateProcessor/UpdatableResource.cs:175]: Same comment as above that the flow and naming is confusing. Maybe you could use the same pattern I suggest above for this section.

File: src/Amazon.Lambda.Tools/TemplateProcessor/UpdatableResource.cs [hunk 5]
```diff
@@ -179,6 +211,11 @@ public string GetMetadataDockerTag()
             {
                 return this.DataSource.GetValueDictionaryFromResource("Metadata", "DockerBuildArgs");
             }
+
+            private bool IsECRImage(string path)
```
Comment (normj) [inline src/Amazon.Lambda.Tools/TemplateProcessor/UpdatableResource.cs:215]: You can make this public and static and then add some unit tests for this method. Since this is a tool and not a library we can be more open to making things public and not worry about changing the API later.

**Review by normj — CHANGES_REQUESTED** [review]

File: src/Amazon.Lambda.Tools/TemplateProcessor/ITemplateParser.cs (hunk 2 above)
Comment (ashishdhingra) [reply src/Amazon.Lambda.Tools/TemplateProcessor/ITemplateParser.cs]: Done.

File: src/Amazon.Lambda.Tools/TemplateProcessor/UpdatableResource.cs (hunk 3 above)
Comment (ashishdhingra) [reply src/Amazon.Lambda.Tools/TemplateProcessor/UpdatableResource.cs]: Done

File: src/Amazon.Lambda.Tools/TemplateProcessor/UpdatableResource.cs (hunk 5 above)
Comment (ashishdhingra) [reply src/Amazon.Lambda.Tools/TemplateProcessor/UpdatableResource.cs]: Done

**Review by normj — APPROVED** [review]

**Review by 96malhar — APPROVED** [review]
